design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,202 @@
+> given failure domain?
+
+> **Open question:** Is it possible to identify a number of OSDs which can be safely updated in
+> parallel without regard to failure domain? For example, try to update some number (or percentage)
+> of OSDs at once in any failure domain. This might be a good default strategy ("best-effort"?).
+
+### Ability to set a maximumum number of OSDs to be updated simultaneously
+In the worst (but unlikely) case, all OSDs which are updated for a given parallel update operation
+might fail to come back online after they are updated. Users may wish to limit the number of OSDs
+updated in parallel in order to avoid too many OSDs failing in this way.
+
+> **Open question:** Is this worth it?
+
+### Cluster growth takes precedence over updates
+Adding new OSDs to a cluster should occur before updating currently-running OSDs. This allows users
+to make use of newly added storage as quickly as possible, which they may need for critical
+applications using the underlying Ceph storage. In some degraded cases, adding new storage may be
+necessary in order to allow currently-running Ceph OSDs to be updated without experiencing storage
+cluster downtime.
+
+This prioritization might delay updates significantly since adding OSDs not only adds capacity to 
+the Ceph cluster but also necessitates data rebalancing. Rebalancing generates data movement which
+needs to settle for updates to be able to proceed.
+
+
+## Technical implementation details
+
+STILL VERY MUCH WIP...
+
+### Definitions
+- **batch**: a list of OSDs that the operator wishes to update simultaneously
+- **query**: unit of work to determine if it is safe to update an OSD
+  - in practice, this is the Ceph command `ceph osd ok-to-stop <osd-id>...`
+- **update**: unit of work to update the OSD Deployment for an OSD
+- **batch query**: unit of work to determine if it is safe to update a batch of OSDs at once
+  - in practice, this is the Ceph command `ceph osd ok-to-stop <osd-id> <osd-id>...`
+- **batch update**: unit of work to update the OSD Deployments for a batch of OSDs at once
+
+### Workflow
+#### High level flow
+1. Build an "update queue" of OSDs in the cluster with failure domain information
+2. Based on the update strategy, find a batch of OSDs from the update queue that can be updated
+3. Remove OSDs in this batch from the update queue
+4. Update the batch of OSD Deployments
+5. If a Deployment fails to be updated, requeue it up to two times, or determine that it failed if
+   this was the third try
+6. If there are remaining OSDs in the update queue, repeat from step 2
+
+#### Build the update queue
+List all OSD Deployments belonging to the Rook cluster to use as the update queue. Go channels are
+likely not the best option since it may be necessary to select individual items from the queue at
+any position and not merely grab the top/bottom item.
+
+Also keep a record of the number of times each OSD has be queued to determine when Rook should stop
```

Comment:
We need to keep trying to update an OSD indefinitely, right? But if you mean that it's a requeue internal to the reconcile, I agree that we would need to cancel and requeue the overall reconcile instead of being stuck inside the same reconcile. If we're in the same reconcile too long, it prevents other actions like updating the cluster CR or even deleting it.

Reply:
I hope to clarify this, but my goal is to retry within the reconcile something like 3-5 times and then exit the reconcile if no progress can be made. We can also specify a reconcile timeout that times out after some period of making no progress to make sure we don't get stuck.